src/devices/src/virtio/block/request.rs

```diff
@@ -224,3 +231,199 @@ impl Request {
+        let supported_request_types = vec![
+            VIRTIO_BLK_T_IN,
+            VIRTIO_BLK_T_OUT,
+            VIRTIO_BLK_T_FLUSH,
+            VIRTIO_BLK_T_GET_ID,
+        ];
+
+        for request_type in supported_request_types {
+            let expected_header = RequestHeader::new(request_type, sector);
+            mem.write_obj_at_addr::<RequestHeader>(expected_header, addr)
+                .unwrap();
+
+            let actual_header = RequestHeader::read_from(&mem, addr).unwrap();
+            assert_eq!(actual_header.request_type, expected_header.request_type);
+            assert_eq!(actual_header.sector, expected_header.sector);
+        }
+
+        // Test that trying to read a request header that goes outside of the
+        // memory boundary fails.
+        assert!(RequestHeader::read_from(&mem, GuestAddress(0x1000)).is_err());
+    }
+
+    #[test]
+    fn test_request_type_from() {
+        assert_eq!(RequestType::from(VIRTIO_BLK_T_IN), RequestType::In);
+        assert_eq!(RequestType::from(VIRTIO_BLK_T_OUT), RequestType::Out);
+        assert_eq!(RequestType::from(VIRTIO_BLK_T_FLUSH), RequestType::Flush);
+        assert_eq!(
+            RequestType::from(VIRTIO_BLK_T_GET_ID),
+            RequestType::GetDeviceID
+        );
+        assert_eq!(RequestType::from(42), RequestType::Unsupported(42));
+    }
+
+    #[test]
+    fn test_execute_error_status() {
+        assert_eq!(
+            ExecuteError::BadRequest(Error::InvalidOffset).status(),
+            VIRTIO_BLK_S_IOERR
+        );
+        assert_eq!(
+            ExecuteError::Flush(io::Error::from_raw_os_error(42)).status(),
+            VIRTIO_BLK_S_IOERR
+        );
+        assert_eq!(
+            ExecuteError::Read(GuestMemoryError::NoMemoryRegions).status(),
+            VIRTIO_BLK_S_IOERR
+        );
+        assert_eq!(
+            ExecuteError::Seek(io::Error::from_raw_os_error(42)).status(),
+            VIRTIO_BLK_S_IOERR
+        );
+        assert_eq!(
+            ExecuteError::Write(GuestMemoryError::NoMemoryRegions).status(),
+            VIRTIO_BLK_S_IOERR
+        );
+        assert_eq!(ExecuteError::Unsupported(42).status(), VIRTIO_BLK_S_UNSUPP);
+    }
+
+    #[test]
+    fn test_parse() {
+        let m = &GuestMemory::new(&[(GuestAddress(0), 0x10000)]).unwrap();
+        let vq = VirtQueue::new(GuestAddress(0), &m, 16);
+
+        assert!(vq.end().0 < 0x1000);
+
+        let request_type_descriptor = 0;
+        let data_descriptor = 1;
+        let status_descriptor = 2;
+
+        vq.avail.ring[0].set(0);
+        vq.avail.idx.set(1);
+
+        {
+            let mut q = vq.create_queue();
+            // Write only request type descriptor.
+            vq.dtable[request_type_descriptor].set(0x1000, 0x1000, VIRTQ_DESC_F_WRITE, 1);
+            let request_header = RequestHeader::new(VIRTIO_BLK_T_OUT, 114);
+            m.write_obj_at_addr::<RequestHeader>(request_header, GuestAddress(0x1000))
```

Comment:
nit: replace this `GuestAddress(0x1000)` with the corresponding descriptor address.